session/tidb.go

```diff
@@ -178,6 +178,17 @@ func runStmt(ctx context.Context, sctx sessionctx.Context, s sqlexec.Statement)
 				history.Count(), sctx.GetSessionVars().IsAutocommit())
 		}
 	}
+	if se.txn.pending() {
+		// After run statement finish, txn state is still pending means the
```

Comment:
So if the statement never needs a Txn(), could we avoid requesting tso from the pd, to save the performance?

Reply:
1. We can not know whether a statement needs Txn() or not before `parse & compile & execute`,
if we don't request `tso` beforehand, then get tso and `parse & compile & execute` can't run in parallel.
2. A tso request will not make heavy impact on pd because, no matter how many requests a batch contains , just one query is sent.